### PR TITLE
Cache service and lots, lots of lots and even more lots of lots of refactoring

### PR DESCRIPTION
**Pull request by renatomassaro**
(Note: depends on #229, which in turn depends on #227)

Specs for the Cache service are discussed at #230.

This PR includes everything proposed on #230, except:

- Inconsistency scanning
- Proactive caching
- Removal of expired entries (currently they are filtered, but not removed)

All of these can be postponed.

It also includes several unrelated refactors

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/helix/231)

### CODE REVIEW

**Comment by renatomassaro** [comment]
Reviewed 29 of 29 files at r3.
Review status: 28 of 72 files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[lib/cache/internal/populate.ex, line 64 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/231#-Kpy451TYdUxZkp8C5TO:-Kpy451TYdUxZkp8C5TP:bhap11t) ([raw file](https://github.com/hackerexperience/helix/blob/8096d9f5f0a5cde878b15acecc09bdf3fb49db1a/lib/cache/internal/populate.ex#L64)):*
> ```Elixir
>   def populate(:storage, storage_id) do
>     with \
>       [drive_id|_] = StorageQuery.get_drive_ids(storage_id),
> ```

I know it sounds ugly, but what do you think of `get_drives_ids`? I think pluralizing the object itself would help clarify the matter.

"get component ids" -> "get components ids"

`components_ids = MotherboardQuery.get_components_ids(motherboard)`
vs
`component_ids = MotherboardQuery.get_component_ids(motherboard)`

---

*[lib/cache/internal/populate.ex, line 114 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/231#-Kpy4-MRAZutHpX8LCGP:-Kpy4-MSTV15HvuW0-JH:b-8fx2u7) ([raw file](https://github.com/hackerexperience/helix/blob/8096d9f5f0a5cde878b15acecc09bdf3fb49db1a/lib/cache/internal/populate.ex#L114)):*
> ```Elixir
>       components_purge = Enum.map(components, &({:component, [&1]}))
> 
>       storages_purge = Enum.map(storage_ids, &({:storage, [&1]}))
> ```

:+1: 

---

*[lib/cache/internal/populate.ex, line 131 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/231#-Kpy3h88CgrOUu9DO4iC:-Kpy3h89g_DiavUIieSA:b-3sig74) ([raw file](https://github.com/hackerexperience/helix/blob/8096d9f5f0a5cde878b15acecc09bdf3fb49db1a/lib/cache/internal/populate.ex#L131)):*
> ```Elixir
> 
>         # Storage
>         Enum.each(storage_ids, &cache(:storage, {&1, server_id}))
> ```

So far I was assuming we'd always have only one storage per motherboard. It's easier to implement. :eyes: 

---

*[lib/cache/model/component_cache.ex, line 38 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/231#-Kpy4pT9dkrbvBPMrQ7w:-Kpy4pT9dkrbvBPMrQ7x:b-7zwvmy) ([raw file](https://github.com/hackerexperience/helix/blob/8096d9f5f0a5cde878b15acecc09bdf3fb49db1a/lib/cache/model/component_cache.ex#L38)):*
> ```Elixir
>     field :motherboard_id, PK
> 
>     field :expiration_date, :utc_datetime
> ```

That's weird, I was using :utc_datetime originally but had some trouble with the resulting Ecto query. But yeah, much better.

---

*[lib/cache/model/server_cache.ex, line 55 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/231#-Kpy6KdY0RTWILd6SNCS:-Kpy6KdY0RTWILd6SNCT:bv6du9q) ([raw file](https://github.com/hackerexperience/helix/blob/8096d9f5f0a5cde878b15acecc09bdf3fb49db1a/lib/cache/model/server_cache.ex#L55)):*
> ```Elixir
>     resources
>     components/a
>   @update_fields ~w/
> ```

Not sure why these models have update params/fields, they are never used. I think that's because they were used before. I'll remove those.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/231)*

**Comment by renatomassaro** [comment]
@mememori, please take a look at the changes at b193962.

Among the usual refactors and cleanups, I've experimentally changed a few things. I'd like your opinion, specially with regards to:

Accidentally I've changed most `$Model.Query` to support both the object's id and the object's type, in which case it would simply pattern match and send the id forward.

Example: 

```elixir
    # Helix.Server.Model.Server.Query
    @spec by_server(Queryable.t, Server.t | Server.id) ::
      Queryable.t
    def by_server(query \\ Server, server_or_server_id)
    def by_server(query, server = %Server{}),
      do: by_server(query, server.server_id)
    def by_server(query, server_id),
      do: where(query, [s], s.server_id == ^server_id)
```

I believe enforcing this behavior *at the model* level brings a lot of flexibility without compromising performance nor readability. This allows us to clean functions like:

from

```elixir
  @spec detach(Server.t | Server.id) ::
    :ok
  def detach(%Server{server_id: id}),
    do: detach(id)
  def detach(server_id) do
    server_id
    |> Server.Query.by_id()
    |> Repo.update_all(set: [motherboard_id: nil])

    :ok
  end
```` 

to

```elixir
  @spec detach(Server.t | Server.id) ::
    :ok
  def detach(server) do
    server
    |> Server.Query.by_server()
    |> Repo.update_all(set: [motherboard_id: nil])

    :ok
  end
```

Our hierarchy "flows" from `Model -> (Internal/State) -> (Query/Action) -> Public -> Driver`, my insight is that it makes sense to be flexible at the lowest level (Model) and gradually tighten the API constraints, according to the Service's requirements.

So, for instance, it may make sense to have Internal accept both IDs and Objects, but Query will only accept IDs. Or, it may make sense to have Query/Action accept both IDs and Objects, but restrain Public to always use IDs.

This is, of course, a suggestion. The drawback is that it adds a new type to the `@spec` and it makes variable names a bit fuzzy[1]. The main advantage is that it *can* put an end at the doubt/inconsistency: "should I use ID or Object?". 

Not sure if you are also annoyed by this inconsistency, but sometimes even within the same *file* I'd have some functions accepting only IDs, and others only Objects.

I don't think this affects tests in neither positive or negative ways:
- Tests that already exists work as usual
- If a new consumer uses a different interface for the first time, we should test the *consumer*.
- If the interface is flawed, the consumer test will fail
- Since most "flexible" interfaces are extremely simple pattern matching, odds of actually doing something wrong (and not being told by Compiler/Dialyzer) are quite small.

Anyway, let me know what you think. What I like about this change is that it's always transparent to the API consumer. So if I change Model to accept both ID and Object, Internal will remain unchanged.

---

[1] So far, we've settled on using either `$thing_id` for ID, or `$thing` for Object. If we do follow this model of sometimes accepting both forms, we'd have to add a third element, identifying that it can be either `$thing_id` or `$thing`. On some parts of the code we use `$thing_or_$thing_id`, but I find it too big. Something like `$think_oid` could be more practical (where `o` stands for `object`)

**Disclaimer**: I've just worked non-stop 12 straight hours. I'm definitely not thinking right, and I'm pretty sure everything I said here is pure garbage. As I said, this refactor was pretty much accidental. Still, there are some cleanups that are worth saving, so I won't trash this commit yet. This also means that I very much probably will not wake up to take you to the bus station. Sorry in advance :man_shrugging:

**Comment by renatomassaro** [comment]
With regards to my last comment, I came up with a few possible solutions.

### Option 1

- Always accept both `t` and `id` on Model.Query, except when not possible (can't think of any scenario)
- Always accept both `t` and `id` on Internal, except when not possible (in some cases it's not possible. For most cases, it is).
- Always accept both `t` and `id` on Query/Action, except when not possible (in most cases it's possible).

Advantage: More flexibility. Slightly improved readability[1].
Disadvantage: Must always look up API documentation (typespec). Slightly harder implementation (must add, sometimes, one or two extra function definitions for pattern matching).

### Option 2

- For service specific data, accept only `t`.
- For external data, accept only `id`.

For instance, `ServerInternal` and `ServerModel` and `ServerQuery` would only work with `Server.t`. External data would be passed as IDs, like `Motherboard.id` or `Entity.id`.

Advantage: only by looking at the module we are calling, we know whether we should send `t` or `id`. When calling ServerQuery.fetch_by_motherboard, we know we should send a `Motherboard.id`.
Disadvantage: Unnecessary (IMO) technical limitation. Slightly reduced readability[1].

### Option 3 (current)

- No standard (use as you want).

Advantage: less hassle to implement.
Disadvantage: Lack of consistency. Must always look up API documentation (typespec). Slightly reduced readability[1].

---

Regardless of the option, tests remains unchanged. I don't think we should aim on testing both types of input (specially since implementation for `t` is usually simply a matter of grabbing the id and passing to the `id` implementation).

[1] One could argue that readability on option 1 is slightly improved, because we could go from

```elixir
wth \
  {:ok, motherboard} <- ServerQuery.get_motherboard(server.server_id)
  {:ok, components} <- MotherboardQuery.get_components(motherboard.motherboard_id)
```

to

```elixir
wth \
  {:ok, motherboard} <- ServerQuery.get_motherboard(server)
  {:ok, components} <- MotherboardQuery.get_components(motherboard)
```

## Variables naming

As I said on the previous comment, regardless of the option we choose, we should add a third sufix for variable, in order to identify whether it refers to `id`, `t` or both, without having to check the typespec.

My proposal:

from

```elixir
motherboard = Motherboard.t
motherboard_id = Motherboard.id
```
to

```elixir
motherboard = Motherboard.t | Motherboard.id
motherboard_id = Motherboard.id
motherboard_t = Motherboard.t
```

Clear, simple, intuitive.

## Method names

Here's a suggestion that I don't think we should go for, but I'll leave it here nonetheless.

On option 1 and 3, calling `ServerQuery.get_resources(?)` begs the question: should I pass an `id` or `t`?

Well, suppose `get_resources` required `t`. Then, we could name it `get_resources_by_t`. If it required `id`, we could name it `get_resources_by_id`. If it accepts both, simply use `get_resources`.

I don't think this is required. Checking the method contract is responsibility of whoever is calling it. On the other hand, if 90%+ of our methods accept both `t` and `id`, then it may make sense to give `id`-specific or `t`-specific methods with a name.

## TL;DR

Please pick option 1, 2 or 3 (or add another one) for input types. I'll then update the code accordingly.

**Comment by renatomassaro** [comment]
(Note: I'll revert the upgrade to Elixir 1.5 and finish it in another PR)

**Comment by mememori** [comment]
Note that there are now some dialyzer warnings on Elixir 1.5.0; those should be fixed on Elixir 1.5.1.

From the Elixir changelog:

> v1.5.1 (2017-08-01)
> ...
> [Kernel] Ensure dialyzer does not emit warnings in some uses of `with`

**Comment by mememori** [comment]
:lgtm:

---

Reviewed 12 of 94 files at r4.
Review status: 5 of 229 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/231#-:-KqyaSH1oKyywlVXYTz1:bnfp4nl)*

**Comment by mememori** [comment]
Reviewed 17 of 62 files at r1, 9 of 24 files at r2, 4 of 29 files at r3, 11 of 94 files at r4, 1 of 30 files at r6, 3 of 36 files at r7, 7 of 59 files at r8, 84 of 129 files at r9, 10 of 31 files at r10, 66 of 75 files at r11, 16 of 16 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/231)*
